crates/wasmtime/src/runtime/vm/sys/unix/vm.rs

```diff
@@ -56,7 +56,7 @@ pub unsafe fn decommit_pages(addr: *mut u8, len: usize) -> io::Result<()> {
                     addr as _,
                     len,
                     ProtFlags::READ | ProtFlags::WRITE,
-                    MapFlags::PRIVATE | MapFlags::FIXED,
+                    MapFlags::PRIVATE | super::mmap::MMAP_NORESERVE_FLAG | MapFlags::FIXED,
```

Comment:
I was wondering about this and the other cases with non-PROT_NONE permissions! From reading the code it wasn't quite clear to me whether this is memory that guests would certainly have accessed and therefore have physical reservations for already, or memory that guests could merely _theoretically_ access.

Reply:
@iximeow actually suggested I do this, so I'm wondering if they have thoughts

Reply:
i may have misremembered where `decommit_pages` gets used! my memory was that this is used when cleaning up after an instance exits, where a new module that could have a smaller max heap size. so my thinking was that even though the pages had been accessible before, we don't yet know if that will still be the case on the instance's next invocation.

Reply:
Heh this function definitely has a storied history and has gone through many iterations. I had to reread up on the callsites to refresh myself on what exactly is happening here. @iximeow you're right that this is used after an instance is deallocated, but this is separate from resizing memory. Here `decommit_pages` is only used to purge backing memory for ranges that were possibly-in-use previously and reset them to their original contents. That's used for both linear memories and tables then. Later other functions like `hide_existing_mapping` are use to shrink memory or `expose_existing_mapping` are used to grow memory.

So put another way: this function is used on memory which was theoretically accessible to wasm but it may or may not have touched because it may not have touched every page in its linear memory. Some of it may be reused for the next instance without further syscalls. Some of it may be hidden via `mprotect` and then re-exposed via `mprotect` to use in wasm later on too.

I suspect Wasmtime's not built currently to work well with precise overcommit because we'd have to hook into `expose_existing_mapping` as well for example which currently is just a bland `mprotect`. Otherwise I don't think it's unreasonable to use `NORESERVE` here.

Overall though it seems like the main use case for `NORESERVE` is in the `reserve` function above where the allocations are often 6G+. Everywhere else should be relatively small allocations/deallocations related to what wasm is about to use or previously using. 

In any case happy to leave this in here and we can always frob it later if necessary.